commandbar.d.ts

```diff
@@ -1,99 +1,125 @@
-// Core primitives
-export type CommandBarContext = Record<string, any>;
-export type CommandBarCallbackArgs = Record<string, any>;
-export type CommandBarCallback = (args: CommandBarCallbackArgs, context: CommandBarContext) => void;
+export interface CommandBar {
+  readonly boot: (args: CommandBarBootArgs) => void;
+  readonly addContext: (context: CommandBarContext) => void;
+  readonly removeContext: (key: string) => void;
+  readonly setContext: (context: CommandBarContext) => void;
+  // K is a type containing types of arguments type K = T1 | T2 | T3
+  readonly addCallback: <K = any>(name: string, callback: CommandBarCallback<K>) => void;
+  readonly removeCallback: (name: string) => void;
+  readonly addCommand: (command: CommandBarCommandConfig) => void;
+  readonly removeCommand: (name: string) => void;
+  readonly addRouter: (routerFunc: CommandBarRouterFunc) => void;
+  readonly addSearch: (category: string, searchFunc: CommandBarSearchFunc) => void;
+  readonly showGuide: (key: string) => void;
+  readonly open: (input?: string) => void;
+  readonly execute: (command: number | string) => void;
+  readonly onboard: () => void;
+  readonly isOpen: () => boolean | undefined;
+  readonly updateContextSettings: <T>(key: string, settings: CommandBarSearchSettings<T>) => void;
+}
 
-// Special cases
-export type CommandBarRouterFunc = (url: string) => void;
-export type CommandBarSearchFunc = (url: string) => void;
+declare global {
+  interface Window {
+    CommandBar: CommandBar;
+  }
+  namespace CommandBarTypes {
+    // Core primitives
+    // If you need several types to be passed to an object
+    // you can pass union type T | K | N  as a generic?
+    export type CommandBarContext<T> = Record<string, T>;
+    export type CommandBarCallbackArgs<T> = Record<string, T>;
+    export type CommandBarCallback<T, K> = (args: CommandBarCallbackArgs<T>, context: CommandBarContext<K>) => void;
 
-// CommandBar.boot
-export type CommandBarBootArgs = {
-    id: string;
-} & Record<string, string | undefined>;
+    // Special cases
+    export type CommandBarRouterFunc = (url: string) => void;
+    export type CommandBarSearchFunc = (url: string) => void;
 
-// Programmatic commands
-export type LinkCommandTemplate = {
-    type: 'link';
-    value: string;
-    operation: 'self' | 'blank' | 'router';
-};
+    // CommandBar.boot
+    export type CommandBarBootArgs = {
+      id: string;
+    } & Record<string, string | undefined>; // Any other arguments? Possible to list them or create an interface?
```

Comment:
Under the hood, `boot` passes its argument to `addContext`, so we can use `CommanadBarContext` here.

Reply:
@jamesaevans Fixed that